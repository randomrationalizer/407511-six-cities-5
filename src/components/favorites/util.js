export const getCitiesFromOffers = (offers) => {
  const cities = offers.slice().map((offer) => offer.city.name).sort();
  return [...new Set(cities)];
};
