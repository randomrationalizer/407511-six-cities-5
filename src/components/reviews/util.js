const REVIEW_MIN_LENGTH = 50;
const REVIEW_MAX_LENGTH = 300;
const RATING_MIN = 1;
const RATING_MAX = 5;

export const checkCommentValidity = (comment) => {
  return comment.length >= REVIEW_MIN_LENGTH && comment.length <= REVIEW_MAX_LENGTH ? true : false;
};

export const checkRatingValidity = (rating) => {
  return rating >= RATING_MIN && rating <= RATING_MAX ? true : false;
};
