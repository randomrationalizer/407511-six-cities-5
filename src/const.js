export const cities = [
  {
    name: `Paris`,
    coords: {
      latitude: 48.864716,
      longitude: 2.349014
    }
  },
  {
    name: `Cologne`,
    coords: {
      latitude: 50.935173,
      longitude: 6.953101
    }
  },
  {
    name: `Brussels`,
    coords: {
      latitude: 50.8505,
      longitude: 4.3488
    }
  },
  {
    name: `Amsterdam`,
    coords: {
      latitude: 52.38333,
      longitude: 4.9
    }
  },
  {
    name: `Hamburg`,
    coords: {
      latitude: 53.551086,
      longitude: 9.993682
    }
  },
  {
    name: `Dusseldorf`,
    coords: {
      latitude: 51.233334,
      longitude: 6.783333
    }
  }
];

export const OfferType = {
  MAIN: `cities`,
  NEARBY: `near-places`,
  FAVORITES: `favorites`
};

export const MapType = {
  MAIN: `cities`,
  PROPERTY: `property`
};

export const SortType = {
  DEFAULT: `DEFAULT`,
  PRICE_UP: `PRICE_UP`,
  PRICE_DOWN: `PRICE_DOWN`,
  RATING_DOWN: `RATING_DOWN`
};
