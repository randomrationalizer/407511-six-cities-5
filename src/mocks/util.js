import configureStore from "redux-mock-store";

export const noop = () => {};

export const getMockStore = (store) => {
  const storeConfig = configureStore();
  return storeConfig(store);
};

export const nope = () => {
  return Promise.resolve({});
};
