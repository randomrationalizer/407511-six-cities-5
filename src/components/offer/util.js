import dayjs from "dayjs";
import {OfferType} from "../../const";


export const capitalize = (str) => {
  return str.charAt(0).toUpperCase() + str.slice(1);
};

// Возвращает дату в виде строки в формате: 2019-04-24
export const formatDate = (date) => {
  return dayjs(date).format(`YYYY-MM-DD`);
};

// Возвращает дату в виде строки в формате: January 2020
export const humanizeDate = (date) => {
  return dayjs(date).format(`MMMM YYYY`);
};

// Сравнение для сортировки отзывов в хронологическом порядке
export const sortByDate = (a, b) => {
  return dayjs(a.date) - dayjs(b.date);
};

export const getDescriptionSentences = (text) => {
  let description = text.split(`. `);
  description[description.length - 1] = description[description.length - 1].replace(`.`, ``);

  return description;
};

export const isFavoritesCard = (type) => {
  return type === OfferType.FAVORITES;
};

export const isMainPageCard = (type) => {
  return type === OfferType.MAIN;
};
