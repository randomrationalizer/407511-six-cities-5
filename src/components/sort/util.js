import {SortType} from "../../const";

export const sortTypeToTitle = {
  [SortType.DEFAULT]: `Popular`,
  [SortType.PRICE_UP]: `Price: low to high`,
  [SortType.PRICE_DOWN]: `Price: high to low`,
  [SortType.RATING_DOWN]: `Top rated first`
};
