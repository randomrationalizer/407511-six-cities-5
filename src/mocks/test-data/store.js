import configureStore from "redux-mock-store";
import {NameSpace} from "../../store/root-reducer";
import mockOffers from "./offers";
import mockReviews from "./reviews";
import mockNearbyOffers from "./nearby-offers";
import mockCurrentOffer from "./current-offer";
import mockUserInfo from "./user-info";
import mockFavorites from "./favorites";
import mockCities from "./cities";
import {SortType, AuthorizationStatus, DEFAULT_CITY} from "../../const";


export default {
  [NameSpace.APP_DATA]: {
    currentCity: DEFAULT_CITY,
    allOffers: mockOffers,
    cities: mockCities,
    currentSort: SortType.DEFAULT
  },
  [NameSpace.USER]: {
    authorizationStatus: AuthorizationStatus.AUTH,
    isAuthRequestComplete: true,
    userInfo: mockUserInfo,
    favorites: mockFavorites
  },
  [NameSpace.CURRENT_OFFER]: {
    offer: mockCurrentOffer,
    reviews: mockReviews,
    nearbyOffers: mockNearbyOffers
  },
  [NameSpace.LOAD_STATUS]: {
    isOffersLoaded: true,
    isFavoritesLoaded: true,
    isCurrentOfferLoaded: true,
    errorMessage: null,
    isLoadFinished: true
  }
};
