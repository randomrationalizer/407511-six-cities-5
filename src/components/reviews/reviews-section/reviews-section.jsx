import React, {useCallback} from "react";
import {connect} from "react-redux";
import {compose} from "redux";
import PropTypes from "prop-types";
import {reviewsPropTypes} from "../review.prop";
import ReviewsList from "../reviews-list/reviews-list";
import ReviewForm from "../review-form/review-form";
import withErrorMessage from "../../../hocs/with-error-message/with-error-message";
import {postNewReview} from "../../../store/api-actions";
import {getOfferReviews} from "../../../store/current-offer/selectors";
import {getAuthorizationStatus} from "../../../store/user/selectors";
import {AuthorizationStatus} from "../../../const";


const ReviewsSection = (props) => {
  const {id, reviews, onFormSubmit, authorizationStatus, showErrorMessage} = props;
  const isAuthorized = authorizationStatus === AuthorizationStatus.AUTH;
  const isAnyReviews = reviews.length ? true : false;

  const handleReviewAdd = useCallback((offerId, newReview) => {
    onFormSubmit(offerId, newReview)
      .catch((error) => showErrorMessage(error.message));
  }, []);

  return (
    <section className="property__reviews reviews">
      <h2 className="reviews__title">
        Reviews &middot; <span className="reviews__amount">{reviews.length}</span>
      </h2>

      {isAnyReviews &&
        <ReviewsList reviews={reviews}/>
      }

      {isAuthorized &&
        <ReviewForm
          key={`${id}-${reviews.length}`}
          onFormSubmit={handleReviewAdd}
          id={id}
        />
      }
    </section>
  );
};

ReviewsSection.propTypes = {
  id: PropTypes.number.isRequired,
  reviews: PropTypes.arrayOf(reviewsPropTypes),
  onFormSubmit: PropTypes.func.isRequired,
  authorizationStatus: PropTypes.string.isRequired,
  showErrorMessage: PropTypes.func.isRequired
};

const mapStateToProps = (state) => ({
  reviews: getOfferReviews(state),
  authorizationStatus: getAuthorizationStatus(state)
});

const mapDispatchToProps = (dispatch) => ({
  onFormSubmit(id, review) {
    return dispatch(postNewReview(id, review));
  }
});

export {ReviewsSection};
export default compose(
    connect(mapStateToProps, mapDispatchToProps),
    withErrorMessage
)(ReviewsSection);
