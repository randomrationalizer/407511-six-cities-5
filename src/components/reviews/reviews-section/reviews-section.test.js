import React from "react";
import renderer from "react-test-renderer";
import {Provider} from "react-redux";
import {ReviewsSection} from "./reviews-section";
import store from "../../../mocks/test-data/store";
import mockReviews from "../../../mocks/test-data/reviews";
import {getMockStore, noop} from "../../../mocks/util";
import {AuthorizationStatus} from "../../../const";


const mockStore = getMockStore(store);


describe(`Should ReviewsSection component renders correctly`, () => {
  describe(`for authorized user`, () => {
    it(`with reviews`, () => {
      const tree = renderer
        .create(
            <Provider store={mockStore}>
              <ReviewsSection
                id={1}
                offerReviews={mockReviews}
                onFormSubmit={noop}
                authorizationStatus={AuthorizationStatus.AUTH}
                setError={noop}
              />
            </Provider>
        ).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it(`without reviews`, () => {
      const tree = renderer
        .create(
            <Provider store={mockStore}>
              <ReviewsSection
                id={1}
                offerReviews={[]}
                onFormSubmit={noop}
                authorizationStatus={AuthorizationStatus.AUTH}
                setError={noop}
              />
            </Provider>
        ).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });

  describe(`for unauthorized user`, () => {
    it(`with reviews`, () => {
      const tree = renderer
        .create(
            <Provider store={mockStore}>
              <ReviewsSection
                id={1}
                offerReviews={mockReviews}
                onFormSubmit={noop}
                authorizationStatus={AuthorizationStatus.NO_AUTH}
                setError={noop}
              />
            </Provider>
        ).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it(`without reviews`, () => {
      const tree = renderer
        .create(
            <Provider store={mockStore}>
              <ReviewsSection
                id={1}
                offerReviews={[]}
                onFormSubmit={noop}
                authorizationStatus={AuthorizationStatus.NO_AUTH}
                setError={noop}
              />
            </Provider>
        ).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
