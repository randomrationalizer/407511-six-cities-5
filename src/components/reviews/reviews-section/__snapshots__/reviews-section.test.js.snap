// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should ReviewsSection component renders correctly for authorized user with reviews 1`] = `
<section
  className="property__reviews reviews"
>
  <h2
    className="reviews__title"
  >
    Reviews · 
    <span
      className="reviews__amount"
    >
      2
    </span>
  </h2>
  <ul
    className="reviews__list"
  >
    <li
      className="reviews__item"
    >
      <div
        className="reviews__user user"
      >
        <div
          className="reviews__avatar-wrapper user__avatar-wrapper"
        >
          <img
            alt="Reviews avatar"
            className="reviews__avatar user__avatar"
            height="54"
            src="https://assets.htmlacademy.ru/intensives/javascript-3/avatar/9.jpg"
            width="54"
          />
        </div>
        <span
          className="reviews__user-name"
        >
          Sophie
        </span>
      </div>
      <div
        className="reviews__info"
      >
        <div
          className="reviews__rating rating"
        >
          <div
            className="reviews__stars rating__stars"
          >
            <span
              style={
                Object {
                  "width": "60%",
                }
              }
            />
            <span
              className="visually-hidden"
            >
              Rating
            </span>
          </div>
        </div>
        <p
          className="reviews__text"
        >
          We loved it so much, the house, the veiw, the location just great.. Highly reccomend :)
        </p>
        <time
          className="reviews__time"
          dateTime="2020-11-02"
        >
          November 2020
        </time>
      </div>
    </li>
    <li
      className="reviews__item"
    >
      <div
        className="reviews__user user"
      >
        <div
          className="reviews__avatar-wrapper user__avatar-wrapper"
        >
          <img
            alt="Reviews avatar"
            className="reviews__avatar user__avatar"
            height="54"
            src="https://assets.htmlacademy.ru/intensives/javascript-3/avatar/7.jpg"
            width="54"
          />
        </div>
        <span
          className="reviews__user-name"
        >
          Mollie
        </span>
      </div>
      <div
        className="reviews__info"
      >
        <div
          className="reviews__rating rating"
        >
          <div
            className="reviews__stars rating__stars"
          >
            <span
              style={
                Object {
                  "width": "40%",
                }
              }
            />
            <span
              className="visually-hidden"
            >
              Rating
            </span>
          </div>
        </div>
        <p
          className="reviews__text"
        >
          What an amazing view! The house is stunning and in an amazing location. The large glass wall had an amazing view of the river!
        </p>
        <time
          className="reviews__time"
          dateTime="2020-11-02"
        >
          November 2020
        </time>
      </div>
    </li>
  </ul>
  <form
    action="#"
    className="reviews__form form"
    method="post"
    onSubmit={[Function]}
  >
    <label
      className="reviews__label form__label"
      htmlFor="review"
    >
      Your review
    </label>
    <div
      className="reviews__rating-form form__rating"
    >
      <input
        checked={false}
        className="form__rating-input visually-hidden"
        id="5-stars"
        name="rating"
        onChange={[Function]}
        required={true}
        type="radio"
        value="5"
      />
      <label
        className="reviews__rating-label form__rating-label"
        htmlFor="5-stars"
        title="perfect"
      >
        <svg
          className="form__star-image"
          height="33"
          width="37"
        >
          <use
            xlinkHref="#icon-star"
          />
        </svg>
      </label>
      <input
        checked={false}
        className="form__rating-input visually-hidden"
        id="4-stars"
        name="rating"
        onChange={[Function]}
        required={true}
        type="radio"
        value="4"
      />
      <label
        className="reviews__rating-label form__rating-label"
        htmlFor="4-stars"
        title="good"
      >
        <svg
          className="form__star-image"
          height="33"
          width="37"
        >
          <use
            xlinkHref="#icon-star"
          />
        </svg>
      </label>
      <input
        checked={false}
        className="form__rating-input visually-hidden"
        id="3-stars"
        name="rating"
        onChange={[Function]}
        required={true}
        type="radio"
        value="3"
      />
      <label
        className="reviews__rating-label form__rating-label"
        htmlFor="3-stars"
        title="not bad"
      >
        <svg
          className="form__star-image"
          height="33"
          width="37"
        >
          <use
            xlinkHref="#icon-star"
          />
        </svg>
      </label>
      <input
        checked={false}
        className="form__rating-input visually-hidden"
        id="2-stars"
        name="rating"
        onChange={[Function]}
        required={true}
        type="radio"
        value="2"
      />
      <label
        className="reviews__rating-label form__rating-label"
        htmlFor="2-stars"
        title="badly"
      >
        <svg
          className="form__star-image"
          height="33"
          width="37"
        >
          <use
            xlinkHref="#icon-star"
          />
        </svg>
      </label>
      <input
        checked={false}
        className="form__rating-input visually-hidden"
        id="1-stars"
        name="rating"
        onChange={[Function]}
        required={true}
        type="radio"
        value="1"
      />
      <label
        className="reviews__rating-label form__rating-label"
        htmlFor="1-stars"
        title="terribly"
      >
        <svg
          className="form__star-image"
          height="33"
          width="37"
        >
          <use
            xlinkHref="#icon-star"
          />
        </svg>
      </label>
    </div>
    <textarea
      className="reviews__textarea form__textarea"
      id="review"
      maxLength={300}
      minLength={50}
      name="review"
      onChange={[Function]}
      placeholder="Tell how was your stay, what you like and what can be improved"
      required={true}
      value=""
    />
    <div
      className="reviews__button-wrapper"
    >
      <p
        className="reviews__help"
      >
        To submit review please make sure to set 
        <span
          className="reviews__star"
        >
          rating
        </span>
         and describe your stay with at least 
        <b
          className="reviews__text-amount"
        >
          50 characters
        </b>
        .
      </p>
      <button
        className="reviews__submit form__submit button"
        disabled={true}
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</section>
`;

exports[`Should ReviewsSection component renders correctly for authorized user without reviews 1`] = `
<section
  className="property__reviews reviews"
>
  <h2
    className="reviews__title"
  >
    Reviews · 
    <span
      className="reviews__amount"
    >
      0
    </span>
  </h2>
  <form
    action="#"
    className="reviews__form form"
    method="post"
    onSubmit={[Function]}
  >
    <label
      className="reviews__label form__label"
      htmlFor="review"
    >
      Your review
    </label>
    <div
      className="reviews__rating-form form__rating"
    >
      <input
        checked={false}
        className="form__rating-input visually-hidden"
        id="5-stars"
        name="rating"
        onChange={[Function]}
        required={true}
        type="radio"
        value="5"
      />
      <label
        className="reviews__rating-label form__rating-label"
        htmlFor="5-stars"
        title="perfect"
      >
        <svg
          className="form__star-image"
          height="33"
          width="37"
        >
          <use
            xlinkHref="#icon-star"
          />
        </svg>
      </label>
      <input
        checked={false}
        className="form__rating-input visually-hidden"
        id="4-stars"
        name="rating"
        onChange={[Function]}
        required={true}
        type="radio"
        value="4"
      />
      <label
        className="reviews__rating-label form__rating-label"
        htmlFor="4-stars"
        title="good"
      >
        <svg
          className="form__star-image"
          height="33"
          width="37"
        >
          <use
            xlinkHref="#icon-star"
          />
        </svg>
      </label>
      <input
        checked={false}
        className="form__rating-input visually-hidden"
        id="3-stars"
        name="rating"
        onChange={[Function]}
        required={true}
        type="radio"
        value="3"
      />
      <label
        className="reviews__rating-label form__rating-label"
        htmlFor="3-stars"
        title="not bad"
      >
        <svg
          className="form__star-image"
          height="33"
          width="37"
        >
          <use
            xlinkHref="#icon-star"
          />
        </svg>
      </label>
      <input
        checked={false}
        className="form__rating-input visually-hidden"
        id="2-stars"
        name="rating"
        onChange={[Function]}
        required={true}
        type="radio"
        value="2"
      />
      <label
        className="reviews__rating-label form__rating-label"
        htmlFor="2-stars"
        title="badly"
      >
        <svg
          className="form__star-image"
          height="33"
          width="37"
        >
          <use
            xlinkHref="#icon-star"
          />
        </svg>
      </label>
      <input
        checked={false}
        className="form__rating-input visually-hidden"
        id="1-stars"
        name="rating"
        onChange={[Function]}
        required={true}
        type="radio"
        value="1"
      />
      <label
        className="reviews__rating-label form__rating-label"
        htmlFor="1-stars"
        title="terribly"
      >
        <svg
          className="form__star-image"
          height="33"
          width="37"
        >
          <use
            xlinkHref="#icon-star"
          />
        </svg>
      </label>
    </div>
    <textarea
      className="reviews__textarea form__textarea"
      id="review"
      maxLength={300}
      minLength={50}
      name="review"
      onChange={[Function]}
      placeholder="Tell how was your stay, what you like and what can be improved"
      required={true}
      value=""
    />
    <div
      className="reviews__button-wrapper"
    >
      <p
        className="reviews__help"
      >
        To submit review please make sure to set 
        <span
          className="reviews__star"
        >
          rating
        </span>
         and describe your stay with at least 
        <b
          className="reviews__text-amount"
        >
          50 characters
        </b>
        .
      </p>
      <button
        className="reviews__submit form__submit button"
        disabled={true}
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</section>
`;

exports[`Should ReviewsSection component renders correctly for unauthorized user with reviews 1`] = `
<section
  className="property__reviews reviews"
>
  <h2
    className="reviews__title"
  >
    Reviews · 
    <span
      className="reviews__amount"
    >
      2
    </span>
  </h2>
  <ul
    className="reviews__list"
  >
    <li
      className="reviews__item"
    >
      <div
        className="reviews__user user"
      >
        <div
          className="reviews__avatar-wrapper user__avatar-wrapper"
        >
          <img
            alt="Reviews avatar"
            className="reviews__avatar user__avatar"
            height="54"
            src="https://assets.htmlacademy.ru/intensives/javascript-3/avatar/9.jpg"
            width="54"
          />
        </div>
        <span
          className="reviews__user-name"
        >
          Sophie
        </span>
      </div>
      <div
        className="reviews__info"
      >
        <div
          className="reviews__rating rating"
        >
          <div
            className="reviews__stars rating__stars"
          >
            <span
              style={
                Object {
                  "width": "60%",
                }
              }
            />
            <span
              className="visually-hidden"
            >
              Rating
            </span>
          </div>
        </div>
        <p
          className="reviews__text"
        >
          We loved it so much, the house, the veiw, the location just great.. Highly reccomend :)
        </p>
        <time
          className="reviews__time"
          dateTime="2020-11-02"
        >
          November 2020
        </time>
      </div>
    </li>
    <li
      className="reviews__item"
    >
      <div
        className="reviews__user user"
      >
        <div
          className="reviews__avatar-wrapper user__avatar-wrapper"
        >
          <img
            alt="Reviews avatar"
            className="reviews__avatar user__avatar"
            height="54"
            src="https://assets.htmlacademy.ru/intensives/javascript-3/avatar/7.jpg"
            width="54"
          />
        </div>
        <span
          className="reviews__user-name"
        >
          Mollie
        </span>
      </div>
      <div
        className="reviews__info"
      >
        <div
          className="reviews__rating rating"
        >
          <div
            className="reviews__stars rating__stars"
          >
            <span
              style={
                Object {
                  "width": "40%",
                }
              }
            />
            <span
              className="visually-hidden"
            >
              Rating
            </span>
          </div>
        </div>
        <p
          className="reviews__text"
        >
          What an amazing view! The house is stunning and in an amazing location. The large glass wall had an amazing view of the river!
        </p>
        <time
          className="reviews__time"
          dateTime="2020-11-02"
        >
          November 2020
        </time>
      </div>
    </li>
  </ul>
</section>
`;

exports[`Should ReviewsSection component renders correctly for unauthorized user without reviews 1`] = `
<section
  className="property__reviews reviews"
>
  <h2
    className="reviews__title"
  >
    Reviews · 
    <span
      className="reviews__amount"
    >
      0
    </span>
  </h2>
</section>
`;
